File: SoundMexPro/tutorial/t_03_error_handling.m
% 
%                    Example for usage of SoundMexPro
% Copyright Hoerzentrum Oldenburg gGmbH Oldenburg 2015, written by Daniel Berg
% 
%    This example shows user defined error handling in SoundMexPro.  
% 
% NOTE: the command 'reseterror' is not handled here. It deals with
% asynchroneous errors occurred in SoundMexPro, i.e. errors, that did not
% occur while executing a SoundMexPro command, but during running
% playback/recording. Such 'asynchroneous errors' are stored and will
% return this error on every command that is called after such an error.
% Handling asynchroneous errors is shown in one of the advanced examples.
% 
% SoundMexPro commands introduced in this example:
%   showerror
%   getlasterror


% call tutorial-initialization script
t_00b_init_tutorial

% initialize SoundMexPro
if 1 ~= soundmexpro('init', ... % command name
        'driver', smpcfg.driver ...  % driver index
        )
    error(['error calling ''init''' error_loc(dbstack)]);
end

smp_disp('standard error handling:');
% set error switch to 'on'
if 1 ~= soundmexpro('showerror', ...    % command name
        'value', 1 ...                  % turn showerror on by value 1
        )
    error(['error calling ''showerror''' error_loc(dbstack)]);
end

% call an unknown. NOTE: we don't check succes flag here, 
% because we want to continue in every case!
soundmexpro('unknown_command');

smp_disp(' ');
smp_disp('user defined error handling:');

% set error switch to 'off'
if 1 ~= soundmexpro('showerror', ...    % command name
        'value', 0 ...                  % turn showerror off by value 0
        )
    error(['error calling ''showerror''' error_loc(dbstack)]);
end

% again call an unknown command
if 1 ~= soundmexpro('unknown_command')
   % retrieve the error string
   [success, original_error] = soundmexpro('getlasterror');
   % this is really bad: error showing is switched off, but 'getlasterror'
   % fails: there's nothing we can do about it ...
   if 1 ~= success
       % display user defined error string...
       error(['error calling ''getlasterror''' error_loc(dbstack)]);
   end
   % now show display user defined error string containig the 'original'
   % error as well
   smp_disp(['An error occurred: ' original_error]);
end

clear soundmexpro;

File: SoundMexPro/tutorialpy/t_03_error_handling.py
#           Example for usage of SoundMexPro
# Copyright Hoerzentrum Oldenburg gGmbH Oldenburg 2021, written by Daniel Berg
#
#       This example shows user defined error handling in SoundMex.  
# 
# Required Python libraries that are not part of the standard library
#   - numpy

import sys
import time
import numpy

# append path to SoundMexPro\bin
sys.path.append('..\\bin')
# import soundmexpro
from soundmexpro import soundmexpro

# NOTE: other than in MATLAB/Octave SoundMexPro raises an error if a command fails, thus
# no need to check a return value after each SoundMexPro command. If you want to change 
# this behaviour please edit ..\bin\soundmexpro.py (look for AssertionError)


# catch an error forced by calling an unknown command

try:
    soundmexpro('unknown')
except:    
    # retrieve the last error ('again': would be part of exception message as well, but 
    # here we want to show hov to retrieve it   
    retvals = soundmexpro('getlasterror')
    # show 'own' error    
    print('error returned from soundmexpro is: {}'.format(retvals['error']))
